File: packages/shared-logging/src/observability.ts
export type LogLevel = "debug" | "info" | "warn" | "error" | "fatal";
export type LogOutcome =
  | "success"
  | "failure"
  | "partial"
  | "denied"
  | "timeout"
  | "cancelled"
  | "skipped";

export interface StructuredLogEnvelope {
  timestamp: string;
  level: LogLevel;
  event: string;
  service: string;
  module?: string;
  environment: string;
  requestId?: string;
  correlationId?: string;
  traceId?: string;
  spanId?: string;
  actorId?: string;
  actorRole?: string;
  mediaId?: string;
  analysisJobId?: string;
  intelligenceJobId?: string;
  campaignJobId?: string;
  attemptId?: string;
  durationMs?: number;
  outcome?: LogOutcome;
  safeErrorCode?: string;
  metadata?: Record<string, unknown>;
}

export interface LogSink {
  readonly sinkId: string;
  emit(event: StructuredLogEnvelope): Promise<void> | void;
}

/** Local/dev console sink that always emits JSON lines. */
export class ConsoleJsonLogSink implements LogSink {
  readonly sinkId = "console-json";
  emit(event: StructuredLogEnvelope): void {
    console.log(JSON.stringify(event));
  }
}

/** Test-only sink used by automated suites. */
export class InMemoryLogSink implements LogSink {
  readonly sinkId = "in-memory";
  readonly events: StructuredLogEnvelope[] = [];
  emit(event: StructuredLogEnvelope): void {
    this.events.push(event);
  }
}

/**
 * OpenTelemetry-compatible exporter abstraction.
 * Automated tests and local default use the in-memory path; production may
 * swap in a real OTLP exporter without changing call sites.
 */
export class OpenTelemetryCompatibleLogSink implements LogSink {
  readonly sinkId = "otel-compatible";
  private readonly delegate: LogSink;
  constructor(delegate: LogSink = new InMemoryLogSink()) {
    this.delegate = delegate;
  }
  emit(event: StructuredLogEnvelope): Promise<void> | void {
    return this.delegate.emit(event);
  }
}

/** Future centralized logging provider adapter (no paid sink required). */
export class ExternalLoggingProviderSink implements LogSink {
  readonly sinkId = "external-provider";
  private readonly delegate: LogSink;
  constructor(delegate: LogSink = new InMemoryLogSink()) {
    this.delegate = delegate;
  }
  emit(event: StructuredLogEnvelope): Promise<void> | void {
    return this.delegate.emit(event);
  }
}

const PRESIGNED_URL_PATTERN =
  /https?:\/\/[^\s"'<>]+(?:X-Amz-Signature|X-Amz-Credential|X-Amz-Algorithm)[^\s"'<>]*/gi;
const AUTHORIZATION_PATTERN = /(Bearer|Basic)\s+[A-Za-z0-9._~+/=-]+/gi;

export function truncateForLog(
  value: string | null | undefined,
  maxChars = 500,
): string | undefined {
  if (!value) {
    return undefined;
  }
  if (value.length <= maxChars) {
    return value;
  }
  return `${value.slice(0, maxChars)}…[truncated]`;
}

export function redactSensitiveText(value: string): string {
  return value
    .replace(PRESIGNED_URL_PATTERN, "[REDACTED_PRESIGNED_URL]")
    .replace(AUTHORIZATION_PATTERN, "[REDACTED_AUTHORIZATION]");
}

const CONTENT_METADATA_KEYS = [
  "ocr",
  "campaign",
  "content",
  "prompt",
  "providerresponse",
  "rawtext",
] as const;

function shouldTruncateMetadataKey(key: string): boolean {
  const normalized = key.toLowerCase().replace(/[_-]/g, "");
  return CONTENT_METADATA_KEYS.some((fragment) => normalized.includes(fragment));
}

export function buildStructuredLog(
  input: Omit<StructuredLogEnvelope, "timestamp"> & {
    timestamp?: string;
  },
): StructuredLogEnvelope {
  const metadata =
    input.metadata === undefined
      ? undefined
      : Object.fromEntries(
          Object.entries(input.metadata).map(([key, value]) => {
            if (typeof value === "string") {
              const redacted = redactSensitiveText(value);
              return [
                key,
                shouldTruncateMetadataKey(key)
                  ? (truncateForLog(redacted, 240) ?? redacted)
                  : redacted,
              ];
            }
            return [key, value];
          }),
        );

  return {
    timestamp: input.timestamp ?? new Date().toISOString(),
    level: input.level,
    event: input.event,
    service: input.service,
    environment: input.environment,
    ...(input.module ? { module: input.module } : {}),
    ...(input.requestId ? { requestId: input.requestId } : {}),
    ...(input.correlationId ? { correlationId: input.correlationId } : {}),
    ...(input.traceId ? { traceId: input.traceId } : {}),
    ...(input.spanId ? { spanId: input.spanId } : {}),
    ...(input.actorId ? { actorId: input.actorId } : {}),
    ...(input.actorRole ? { actorRole: input.actorRole } : {}),
    ...(input.mediaId ? { mediaId: input.mediaId } : {}),
    ...(input.analysisJobId ? { analysisJobId: input.analysisJobId } : {}),
    ...(input.intelligenceJobId ? { intelligenceJobId: input.intelligenceJobId } : {}),
    ...(input.campaignJobId ? { campaignJobId: input.campaignJobId } : {}),
    ...(input.attemptId ? { attemptId: input.attemptId } : {}),
    ...(input.durationMs !== undefined ? { durationMs: input.durationMs } : {}),
    ...(input.outcome ? { outcome: input.outcome } : {}),
    ...(input.safeErrorCode ? { safeErrorCode: input.safeErrorCode } : {}),
    ...(metadata ? { metadata } : {}),
  };
}

export function withTraceContext(
  parent: Pick<StructuredLogEnvelope, "traceId" | "spanId" | "correlationId" | "requestId">,
  child: Omit<StructuredLogEnvelope, "timestamp">,
): StructuredLogEnvelope {
  const traceId = child.traceId ?? parent.traceId ?? parent.correlationId ?? parent.requestId;
  const correlationId = child.correlationId ?? parent.correlationId ?? parent.requestId;
  const requestId = child.requestId ?? parent.requestId;
  return buildStructuredLog({
    ...child,
    ...(traceId ? { traceId } : {}),
    ...(child.spanId ? { spanId: child.spanId } : {}),
    ...(correlationId ? { correlationId } : {}),
    ...(requestId ? { requestId } : {}),
  });
}

File: apps/ai-service/app/core/observability.py
"""Shared structured log envelope helpers for the FastAPI AI service."""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

from app.core.logging import redact_value

CONTENT_METADATA_KEYS = ("ocr", "campaign", "content", "prompt", "providerresponse", "rawtext")


def truncate_for_log(value: str, max_chars: int = 240) -> str:
    if len(value) <= max_chars:
        return value
    return f"{value[:max_chars]}…[truncated]"


def _should_truncate_metadata_key(key: str) -> bool:
    normalized = key.lower().replace("_", "").replace("-", "")
    return any(fragment in normalized for fragment in CONTENT_METADATA_KEYS)


def build_structured_log(
    *,
    level: str,
    event: str,
    service: str,
    environment: str,
    module: str | None = None,
    request_id: str | None = None,
    correlation_id: str | None = None,
    trace_id: str | None = None,
    span_id: str | None = None,
    actor_id: str | None = None,
    actor_role: str | None = None,
    media_id: str | None = None,
    analysis_job_id: str | None = None,
    intelligence_job_id: str | None = None,
    campaign_job_id: str | None = None,
    attempt_id: str | None = None,
    duration_ms: int | None = None,
    outcome: str | None = None,
    safe_error_code: str | None = None,
    metadata: dict[str, Any] | None = None,
) -> dict[str, Any]:
    sanitized_metadata: dict[str, Any] | None = None
    if metadata is not None:
        sanitized_metadata = {}
        for key, value in metadata.items():
            redacted = redact_value(key, value)
            if isinstance(redacted, str) and _should_truncate_metadata_key(key):
                sanitized_metadata[key] = truncate_for_log(redacted)
            else:
                sanitized_metadata[key] = redacted

    envelope: dict[str, Any] = {
        "timestamp": datetime.now(UTC).isoformat(),
        "level": level,
        "event": event,
        "service": service,
        "environment": environment,
    }
    optional = {
        "module": module,
        "requestId": request_id,
        "correlationId": correlation_id,
        "traceId": trace_id,
        "spanId": span_id,
        "actorId": actor_id,
        "actorRole": actor_role,
        "mediaId": media_id,
        "analysisJobId": analysis_job_id,
        "intelligenceJobId": intelligence_job_id,
        "campaignJobId": campaign_job_id,
        "attemptId": attempt_id,
        "durationMs": duration_ms,
        "outcome": outcome,
        "safeErrorCode": safe_error_code,
        "metadata": sanitized_metadata,
    }
    for key, value in optional.items():
        if value is not None:
            envelope[key] = value
    return envelope
